### PR TITLE
[Scheduler] Add new component documentation

### LINKED ISSUE #18067: [Messenger] Add a scheduler component
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#47112
| PR author(s) | @upyx, @fabpot
| Merged in    | 6.3

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 6.3 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smile:

### PR DESCRIPTION
**Pull request by MrYamous**
Fixes #18067

This is an early draft for the Scheduler component and his Messenger integration to gather early feedback

While starting the documentation, I asked myself a few questions :

- Scheduler is a standalone component but shares some keys concepts with Messenger (message, handlers...), should we document them here too or only refers to related parts in Messenger doc ? 
- There are already several issues in docs for different parts of the component (initial integration, the debug command, new triggers...), should we have them all in the same pull request or keep split ?

### CODE REVIEW

File: scheduler.rst [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    use Symfony\Component\Scheduler\Schedule;
+    use Symfony\Component\Scheduler\ScheduleProviderInterface;
+
+    #[AsSchedule('trial')]
+    final class EndOfTrialScheduleProvider implements ScheduleProviderInterface
+    {
+        public function getSchedule(string $id): Schedule
+        {
+            return (new Schedule())
+                ->add(new RecurringMessage::every('1 week', new EndofTrialMessage()))
```
Comment (quentint) [inline scheduler.rst:47]: The `new` keyword should not be there 😉 

```suggestion
                ->add(RecurringMessage::every('1 week', new EndofTrialMessage()))
```

File: scheduler.rst [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    $ composer require symfony/scheduler
+
+.. include:: /components/require_autoload.rst.inc
+
+.. _register-schedule-provider:
+
+Register a Schedule provider
+----------------------------
+
+A Schedule provider is a class defining your schedule : which messages should
```
Comment (kbond) [inline scheduler.rst:30]:
```suggestion
A Schedule provider is a class defining your schedule. Which messages should 
```

File: scheduler.rst [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+.. include:: /components/require_autoload.rst.inc
+
+.. _register-schedule-provider:
+
+Register a Schedule provider
+----------------------------
+
+A Schedule provider is a class defining your schedule : which messages should 
+be processed depending a trigger you choose::
```
Comment (kbond) [inline scheduler.rst:31]:
```suggestion
be processed depends on the trigger you choose::
```

File: scheduler.rst [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    namespace App\Schedule;
+
+    use App\Message\EndofTrialMessage;
+    use Symfony\Component\Scheduler\Attribute\AsSchedule;
+    use Symfony\Component\Scheduler\RecurringMessage;
+    use Symfony\Component\Scheduler\Schedule;
+    use Symfony\Component\Scheduler\ScheduleProviderInterface;
+
+    #[AsSchedule('trial')]
```
Comment (kbond) [inline scheduler.rst:41]:
```suggestion
    #[AsSchedule('default')]
```

I'm thinking we should just use default here. Multiple schedules is an advanced use-case imo.

File: scheduler.rst [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    final class EndOfTrialScheduleProvider implements ScheduleProviderInterface
```
Comment (kbond) [inline scheduler.rst:42]:
```suggestion
    final class AppSchedule implements ScheduleProviderInterface
```

File: scheduler.rst [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+--------------------------
+
+Scheduler uses the same :ref:`Messenger worker <messenger-worker>` to consume
+messages dispatched by your schedule class.
+
+You can do this with the ``messenger:consume`` command:
+
+.. code-block:: terminal
+
+    $ php bin/console messenger:consume scheduler_trial
```
Comment (kbond) [inline scheduler.rst:62]:
```suggestion
    $ php bin/console messenger:consume scheduler_default
```

File: scheduler.rst [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+Trigger your schedule
+---------------------
+
+Symfony provides by default two triggers for your schedules : `every` and
+`cron`. By extending the provided TriggerInterface you can also create
+your own trigger for your app.
```
Comment (kbond) [inline scheduler.rst:69]:
```suggestion
Symfony provides several triggers, and trigger wrappers out of the box. You can also define your own triggers.
```

File: scheduler.rst [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+Every
```
Comment (kbond) [inline scheduler.rst:71]:
```suggestion
PeriodicalTrigger
```

File: scheduler.rst [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    // using a DateTimeImmutable object
+    $from = new \DateTimeImmutable('13:47', new \DateTimeZone('Europe/Paris'));
+    RecurringMessage::every('first day of month', $msg, from: $from);
+
+If your schedule do not need to run indefinitely, you can use `until` option
+to let the Component know when the trigger should stop::
+
+    RecurringMessage::every('monday', $msg, until: '2023-06-12');
+
+Cron
```
Comment (kbond) [inline scheduler.rst:101]:
```suggestion
CronExpressionTrigger
```

File: scheduler.rst [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+~~~~
+
```
Comment (kbond) [inline scheduler.rst:103]: We'll need a note that https://github.com/dragonmantank/cron-expression is required to use this trigger.

File: scheduler.rst [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+This trigger can be configured following same rules as the eponymous Unix
+utility::
+
+    // trigger every hour
+    RecurringMessage::cron('0 * * * *', $msg);
+
+    // trigger every monday at 12:00
+    RecurringMessage::cron('0 12 * * 1', $msg);
+
+.. note::
```
Comment (kbond) [inline scheduler.rst:113]: We should document https://github.com/symfony/symfony/pull/49792

File: scheduler.rst [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+    The minimal interval allowed with cron is 1 minute.
+
```
Comment (kbond) [inline scheduler.rst:116]: TODO:

- JitterTrigger
- CallbackTrigger
- ExcludeTimeTrigger

**Review by kbond — COMMENTED** [review]
Excellent start!

File: scheduler.rst (hunk 3 above)
Comment (Curryed) [reply scheduler.rst:31]:
```suggestion
be processed depending on the trigger you choose:
```

**Comment by alexander-schranz** [comment]
Is there something we can help here to move this forward? Think at current state would be better to have a incomplete documentation then no documentation of the scheduler component.

**Comment by MrYamous** [comment]
Sorry I created this PR then I've been pretty absent.
AFAIK Scheduler is still undocumented, but it's on hold referring to #18290 ([see comment](https://github.com/symfony/symfony-docs/pull/18290#issuecomment-1700535076))

@javiereguiluz Maybe we can help in some way ?

**Comment by MrYamous** [comment]
Closing in favor of #19244

**Pull request closed without merging**